crates/next-core/src/manifest.rs

```diff
@@ -0,0 +1,103 @@
+    ContentSourceContent, ContentSourceData, ContentSourceResultVc, ContentSourceVc,
+};
+use turbopack_node::{
+    node_api_source::NodeApiContentSourceVc, node_rendered_source::NodeRenderContentSourceVc,
+};
+
+#[turbo_tasks::value(shared)]
+pub struct DevManifestContentSource {
+    pub page_roots: Vec<ContentSourceVc>,
+}
+
+#[turbo_tasks::value_impl]
+impl DevManifestContentSourceVc {
+    #[turbo_tasks::function]
+    async fn find_routes(self) -> Result<StringsVc> {
+        let this = &*self.await?;
+        let mut queue = this.page_roots.clone();
+        let mut routes = IndexSet::new();
+
+        while let Some(content_source) = queue.pop() {
+            if let Some(combined_source) =
+                CombinedContentSourceVc::resolve_from(content_source).await?
+            {
+                queue.extend(combined_source.await?.sources.iter().copied());
+
+                continue;
+            }
+
+            if let Some(conditional_source) =
+                ConditionalContentSourceVc::resolve_from(content_source).await?
+            {
+                let conditional_source = &*conditional_source.await?;
+                queue.push(conditional_source.activator);
+                queue.push(conditional_source.action);
+
+                continue;
+            }
+
+            if let Some(api_source) = NodeApiContentSourceVc::resolve_from(content_source).await? {
+                routes.insert(format!("/{}", api_source.get_pathname().await?));
+
+                continue;
+            }
+
+            if let Some(page_source) =
+                NodeRenderContentSourceVc::resolve_from(content_source).await?
+            {
+                routes.insert(format!("/{}", page_source.get_pathname().await?));
+
+                continue;
+            }
+
+            // ignore anything else
+        }
+
+        routes.sort();
+
+        Ok(StringsVc::cell(routes.into_iter().collect()))
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl ContentSource for DevManifestContentSource {
+    #[turbo_tasks::function]
+    async fn get(
+        self_vc: DevManifestContentSourceVc,
+        path: &str,
+        _data: turbo_tasks::Value<ContentSourceData>,
+    ) -> Result<ContentSourceResultVc> {
+        let requested_manifest = match path.rsplit_once('/') {
+            Some(("_next/static/development", file))
+                if file == "_devPagesManifest.json" || file == "_devMiddlewareManifest.json" =>
+            {
+                file
+            }
+            _ => return Ok(ContentSourceResultVc::not_found()),
+        };
+
+        let content = if requested_manifest == "_devPagesManifest.json" {
+            let pages = &*self_vc.find_routes().await?;
+
+            serde_json::to_string(&serde_json::json!({
+                "pages": pages,
+            }))?
+        } else {
+            // empty middleware manifest
+            "[]".to_string()
+        };
```

Comment:
You could do this in the match above instead if you separate the "_devPagesManifest.json" and "_devMiddlewareManifest.json" cases.